### ISSUE TITLE
Focal Loss

### ISSUE THREAD

**Issue by sebastian-sz** [keras-team/keras-cv#319]
Focal loss was used to train [RetinaNet](https://arxiv.org/pdf/1708.02002v2.pdf).

From [Papers with code](https://paperswithcode.com/method/focal-loss):
> A Focal Loss function addresses class imbalance during training in tasks like object detection.

Reference Keras Implementation:
https://keras.io/examples/vision/retinanet/#implementing-smooth-l1-loss-and-focal-loss-as-keras-custom-losses 

What do you think about this component? I think that maybe it could also benefit other models.

**Comment by innat** [comment]
sigmoid focal loss : https://www.tensorflow.org/addons/api_docs/python/tfa/losses/SigmoidFocalCrossEntropy

**Comment by bhack** [comment]
https://discuss.tensorflow.org/t/focal-loss-function-for-multilabel-classification/8933

**Comment by sebastian-sz** [comment]
@bhack I understand that we would like to avoid duplication across repositories. I did not know there is an implementation already in Keras core (although it lacks the `alpha` parameter). 

Should I close this issue?

**Comment by bhack** [comment]
> @bhack I understand that we would like to avoid duplication across repositories. I did not know there is an implementation already in Keras core (although it lacks the `alpha` parameter). 
> 
> Should I close this issue?

Personally I am ok to migrate things here.

**Comment by sebastian-sz** [comment]
@bhack Ok thanks.

**Comment by bhack** [comment]
@sebastian-sz Honestly It was the scope of https://github.com/keras-team/keras-cv/issues/74 to do this in an orderly manner but it seems that in the end it is quite recurrent to never manage this kind of things in a clean way.

**Comment by LukeWood** [comment]
> @bhack I understand that we would like to avoid duplication across repositories. I did not know there is an implementation already in Keras core (although it lacks the `alpha` parameter).
> 
> Should I close this issue?

No, this is a good issue to have open.  TFA is not the same as the Keras API, even if some of the objects are in fact Keras `Layer` objects.  TFA is a great place to look at reference implementations, but we should not expect users to install TFA to use KerasCV.  Users should be able to coherently use KerasCV end to end using only KerasCV, Keras core, and TensorFlow core components (i.e. tf.data).

**Comment by quantumalaviya** [comment]
@LukeWood I can work on this!

**Comment by innat** [comment]
prev implementation on kcv
https://github.com/keras-team/keras-cv/blob/98eae129b6795bb4bfbc888e9e7e916f5d418da3/keras_cv/losses/focal_loss.py

**Comment by SamuelMarks** [comment]
@innat I just changed the docstring… @tanzhenyu did the rest I think
